src/Glimpse.Server.Web/Storage/InMemoryStorage.cs

```diff
@@ -8,65 +7,200 @@ namespace Glimpse.Server.Web
 {
     public class InMemoryStorage : IStorage, IQueryRequests
     {
+        /// <summary>
+        /// Internal class to contain all data associated with a single request. 
+        /// </summary>
+        private class RequestInfo
+        {
+            public RequestInfo(LinkedListNode<Guid> requestIdNode)
+            {
+                this.RequestIdNode = requestIdNode;
+                this.Messages = new LinkedList<IMessage>();
+                this.Indices = null;
+            }
+
+            /// <summary>
+            /// List of individual messages associated with this request
+            /// </summary>
+            public LinkedList<IMessage> Messages { get; private set; }
+            
+            /// <summary>
+            /// The node in the _activeRequests list.  Storing this here allows us to determine to track 
+            /// requests by age in constant time.
+            /// </summary>
+            public LinkedListNode<Guid> RequestIdNode { get; private set; }
+
+            /// <summary>
+            /// Indices associated with this Request
+            /// </summary>
+            public RequestIndices Indices { get; set; }
+        }
+
         public const int RequestsPerPage = 50;
-        public const int MaxRequests = 500; // TODO: Make this configurable
+        public const int DefaultMaxRequests = 500;
 
-        private readonly List<IMessage> _messages;
-        private readonly ConcurrentDictionary<Guid, RequestIndices> _indices;
+        /// <summary>
+        /// Primary storage for Messages. 
+        /// </summary>
+        private readonly Dictionary<Guid, RequestInfo> _requestTable;
 
-        private static readonly object _locker = new object(); // HACK :(
+        /// <summary>
+        /// Currently active requests, ordered by incoming messages for the request.  The first entry in the list will be
+        /// the most recent request to receive a message, the last entry the oldest request to receive a message.
+        /// </summary>
+        private readonly LinkedList<Guid> _activeRequests;
 
-        public InMemoryStorage()
+        /// <summary>
+        /// Maximum number of requests to store for this request.
+        /// </summary>
+        private readonly int _maxRequests;
+
+
+        /// <summary>
+        /// Lock to synchronize access to this data structure. 
+        /// </summary>
+        // TODO:  Consider changing to a multi-reader/single-writer lock for increased read throughput
+        private static readonly object _locker = new object();
```

Comment:
Fixed.

From: Anthony van der Hoorn [mailto:notifications@github.com]
Sent: Friday, September 25, 2015 10:13 AM
To: Glimpse/Glimpse.Prototype <Glimpse.Prototype@noreply.github.com>
Cc: Mike Kaufman <Mike.Kaufman@microsoft.com>
Subject: Re: [Glimpse.Prototype] 55 in memory storage datastructures (#62)


In src/Glimpse.Server.Web/Storage/InMemoryStorage.cs<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2fGlimpse%2fGlimpse.Prototype%2fpull%2f62%23discussion_r40452665&data=01%7c01%7cmike.kaufman%40microsoft.com%7c3050e276bd83428cc51808d2c5cc9274%7c72f988bf86f141af91ab2d7cd011db47%7c1&sdata=ogGSwMYr1p%2fyiMK4iJsj8wh3KtkH5LiZ2qDzVGVv62w%3d>:

>

> -        public InMemoryStorage()

> +        /// <summary>

> +        /// Maximum number of requests to store for this request.

> +        /// </summary>

> +        private readonly int _maxRequests;

> +

> +

> +        /// <summary>

> +        /// Lock to synchronize access to this data structure.

> +        /// </summary>

> +        // TODO:  Consider changing to a multi-reader/single-writer lock for increased read throughput

> +        private static readonly object _locker = new object();

Does this need to be static?

—
Reply to this email directly or view it on GitHub<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2fGlimpse%2fGlimpse.Prototype%2fpull%2f62%2ffiles%23r40452665&data=01%7c01%7cmike.kaufman%40microsoft.com%7c3050e276bd83428cc51808d2c5cc9274%7c72f988bf86f141af91ab2d7cd011db47%7c1&sdata=pNulmOwQip6rXfOc5wWtGxNZtkN7pvuH%2biTFL6PkWZk%3d>.